### ISSUE TITLE
The model's hidden state is inconsistent when using padding.

### ISSUE THREAD

**Issue by hhhhhy60** [QwenLM/Qwen2#829]
experiment on qwen2 7b instruct, config like this:

```
model_path = "qwen2-7b-instruct"

model = AutoModelForCausalLM.from_pretrained(
    model_path,
    torch_dtype="auto",
    device_map="cuda:1"
)
model.eval()
tokenizer = AutoTokenizer.from_pretrained(model_path,padding_side='left')
```

here is the experiment:

```
torch.set_printoptions(precision=32, threshold=8, edgeitems=2, linewidth=80, sci_mode=False)

batch = ['你好啊']

batch_message_list = [[{'role': 'user', 'content': x}] for x in batch]

batch_template = [tokenizer.apply_chat_template(
      conversation = message_list,
      tokenize=False,
      add_generation_prompt=True
    ) for message_list in batch_message_list]

input_ = tokenizer(batch_template, padding='longest',return_tensors='pt',).to(device)

with torch.no_grad():
    output = model.forward(**input_,output_hidden_states=True)
    
print(output.hidden_states[-1][0][-1])
```

output:
```
tensor([-2.03125000000000000000000000000000,
         1.35937500000000000000000000000000,
         ...,
        -1.57812500000000000000000000000000,
         0.74609375000000000000000000000000], device='cuda:1',
       dtype=torch.bfloat16)
```

then manual adding one padding token, i think is same with auto padding result by tokenizer
```
torch.set_printoptions(precision=32, threshold=8, edgeitems=2, linewidth=80, sci_mode=False)


batch = ['你好啊']

batch_message_list = [[{'role': 'user', 'content': x}] for x in batch]

batch_template = [tokenizer.apply_chat_template(
      conversation = message_list,
      tokenize=False,
      add_generation_prompt=True
    ) for message_list in batch_message_list]

input_ = tokenizer(batch_template, padding='longest',return_tensors='pt',).to(device)


input_['input_ids'] = torch.concat((torch.tensor([[tokenizer.pad_token_id]],device=device),input_['input_ids']), dim=1)
input_['attention_mask'] = torch.concat((torch.tensor([[0]],device=device),input_['attention_mask']), dim=1)


with torch.no_grad():
    output = model.forward(**input_,output_hidden_states=True)
    
print(output.hidden_states[-1][0][-1])
```

output:
```
tensor([-2.01562500000000000000000000000000,
         1.34375000000000000000000000000000,
         ...,
        -1.69531250000000000000000000000000,
         0.65625000000000000000000000000000], device='cuda:1',
       dtype=torch.bfloat16)
```

the hidden state is different, which makes batch generate result diff to single sample generation.
